gamemodes/terrortown/gamemode/shared/sh_marker_vision_object.lua

```diff
@@ -0,0 +1,244 @@
+-- @param Entity owner The mark owner
+-- @realm shared
+function MARKER_VISION_OBJECT:SetOwner(owner)
+	self.data.owner = owner
+end
+
+---
+-- Gets the mark owner to the element.
+-- @return Entity The mark owner
+-- @realm shared
+function MARKER_VISION_OBJECT:GetOwner()
+	return self.data.owner
+end
+
+---
+-- Sets the visible for flag to the element.
+-- @param number visibleFor The visible for flag
+-- @realm shared
+function MARKER_VISION_OBJECT:SetVisibleFor(visibleFor)
+	self.data.visibleFor = visibleFor
+end
+
+---
+-- Gets the visible for flag to the element.
+-- @return number The visible for flag
+-- @realm shared
+function MARKER_VISION_OBJECT:GetVisibleFor()
+	return self.data.visibleFor
+end
+
+---
+-- Gets the visible for flag as a language string to the element.
+-- @return string The visible for flag as a language string
+-- @realm shared
+function MARKER_VISION_OBJECT:GetVisibleForString()
+	return "marker_vision_visible_for_" .. tostring(self:GetVisibleFor())
+end
+
+---
+-- Checks if this element is the searched object defined by the unique
+-- identifier and the entity that is marked.
+-- @param Entity ent The entitiy that is possibly marked
+-- @param string identifier The unique identifier
+-- @return boolean Returns true if the element matches
+-- @realm shared
+function MARKER_VISION_OBJECT:IsSearchedObject(ent, identifier)
+	return self.data.ent == ent and self.data.identifier == identifier
+end
+
+---
+-- Updates the relations of this marker vision element.
+-- @note This does not sync it to the client, call @{MARKER_VISION_OBJECT:SyncToClients} to sync to client.
+-- @param number|string|Player owner The owner of the wallhack that takes their ownership with them on
+-- team change; can also be a team (string) or role (number) if it shouldn't be bound to a player
+-- @param number visibleFor Visibility setting: `VISIBLE_FOR_PLAYER`, `VISIBLE_FOR_ROLE`, `VISIBLE_FOR_TEAM`,
+-- `VISIBLE_FOR_ALL`
+-- @realm shared
+function MARKER_VISION_OBJECT:UpdateRelations(owner, visibleFor)
+	-- this is done this way to support an UpdateRelations call on Role/Team change without
+	-- the need to pass these values again
+	self.data.owner = owner or self.data.owner
+	self.data.visibleFor = visibleFor or self.data.visibleFor
+
+	if SERVER then
+		if self.data.visibleFor == VISIBLE_FOR_PLAYER then
+			self.receiverList = {self.data.owner}
+		elseif self.data.visibleFor == VISIBLE_FOR_ROLE then
+			if IsPlayer(self.data.owner) then
+				self.receiverList = GetRoleFilter(self.data.owner:GetSubRole(), false)
+			else
+				-- handle static role
+				self.receiverList = GetRoleFilter(self.data.owner, false)
+			end
+		elseif self.data.visibleFor == VISIBLE_FOR_TEAM then
+			if IsPlayer(self.data.owner) then
+				self.receiverList = GetTeamFilter(self.data.owner:GetTeam(), false, true)
+			else
+				-- handle static team
+				self.receiverList = GetTeamFilter(self.data.owner, false, true)
+			end
+		else
+			self.receiverList = nil
```

Comment:
I see what you mean but it is expected to call this function and the sync function together. I thought it looked cleaner to split it up

Reply:
I think you missed my point here again. This is more about keeping these lists in sync:
Now you need to update every marker vision object when a player leaves or connects etc. and the playerlist and thus the potential receivers change. This can easily be avoided by not precomputing these and moving this to the latest possible computation, where it is actually needed.
It just makes it easier to not accidentally miss updating the receivers.